File: sdks/py/src/jeffs_brain_memory/retrieval/rrf.py
# SPDX-License-Identifier: Apache-2.0
"""Reciprocal Rank Fusion (Cormack, Clarke, Buettcher 2009).

Merges an arbitrary number of ranked lists into a single ranking with
``rrf_score(doc) = sum over lists of 1 / (k + rank(doc) + 1)``. Order
of the input lists does not change the fused score; it only controls
which list seeds the metadata for a candidate seen in multiple lists.

Mirrors ``sdks/go/retrieval/rrf.go`` bit for bit.
"""

from __future__ import annotations

from dataclasses import dataclass

from .types import RetrievedChunk

RRF_DEFAULT_K = 60
"""Canonical constant. The spec pins k = 60 across every SDK."""


@dataclass(slots=True)
class RRFCandidate:
    """Input shape accepted by :func:`reciprocal_rank_fusion`."""

    id: str = ""
    path: str = ""
    title: str = ""
    summary: str = ""
    content: str = ""
    bm25_rank: int = 0
    have_bm25_rank: bool = False
    vector_similarity: float = 0.0
    have_vector_sim: bool = False


@dataclass(slots=True)
class _Bucket:
    id: str
    path: str
    title: str
    summary: str
    content: str
    bm25_rank: int
    have_bm25_rank: bool
    vector_similarity: float
    have_vector_sim: bool
    score: float


def reciprocal_rank_fusion(
    lists: list[list[RRFCandidate]], k: int = RRF_DEFAULT_K
) -> list[RetrievedChunk]:
    """Fuse ranked lists into a single ranking.

    The one-way fill rule from the spec: later lists seed metadata
    (title/summary/content/bm25_rank/vector_similarity) only when the
    first-seen entry left them empty. Ties on score break by path
    ascending for stable output.
    """
    safe_k = k if k > 0 else RRF_DEFAULT_K

    buckets: dict[str, _Bucket] = {}
    order: list[str] = []

    for candidates in lists:
        for rank, cand in enumerate(candidates):
            if not cand.id:
                continue
            contribution = 1.0 / float(safe_k + rank + 1)
            existing = buckets.get(cand.id)
            if existing is None:
                buckets[cand.id] = _Bucket(
                    id=cand.id,
                    path=cand.path,
                    title=cand.title,
                    summary=cand.summary,
                    content=cand.content,
                    bm25_rank=cand.bm25_rank if cand.have_bm25_rank else 0,
                    have_bm25_rank=cand.have_bm25_rank,
                    vector_similarity=(
                        cand.vector_similarity if cand.have_vector_sim else 0.0
                    ),
                    have_vector_sim=cand.have_vector_sim,
                    score=contribution,
                )
                order.append(cand.id)
                continue

            if not existing.title and cand.title:
                existing.title = cand.title
            if not existing.summary and cand.summary:
                existing.summary = cand.summary
            if not existing.content and cand.content:
                existing.content = cand.content
            if not existing.have_bm25_rank and cand.have_bm25_rank:
                existing.bm25_rank = cand.bm25_rank
                existing.have_bm25_rank = True
            if not existing.have_vector_sim and cand.have_vector_sim:
                existing.vector_similarity = cand.vector_similarity
                existing.have_vector_sim = True
            existing.score += contribution

    out: list[RetrievedChunk] = []
    for id_ in order:
        b = buckets[id_]
        chunk = RetrievedChunk(
            chunk_id=b.id,
            document_id=b.id,
            path=b.path,
            score=b.score,
            text=b.content,
            title=b.title,
            summary=b.summary,
        )
        if b.have_bm25_rank:
            chunk.bm25_rank = b.bm25_rank
        if b.have_vector_sim:
            chunk.vector_similarity = b.vector_similarity
        out.append(chunk)

    # Stable sort: score desc, then path asc.
    out.sort(key=lambda c: (-c.score, c.path))
    return out

File: go/retrieval/rrf.go
// SPDX-License-Identifier: Apache-2.0

package retrieval

import "sort"

// RRFDefaultK is the canonical constant in the Reciprocal Rank Fusion
// formula: rrf_score(doc) = sum over lists of 1 / (k + rank(doc)).
// Cormack, Clarke, Buettcher (SIGIR 2009) tested k in [10, 1000] and
// found k = 60 robust across TREC tracks; the spec pins it at 60.
const RRFDefaultK = 60

// rrfCandidate is the shape the fusion step accepts. It mirrors the
// TypeScript RRFCandidate so callers that port tests from the TS
// reference can keep the same field layout.
type rrfCandidate struct {
	id               string
	path             string
	title            string
	summary          string
	content          string
	bm25Rank         int
	haveBM25Rank     bool
	vectorSimilarity float64
	haveVectorSim    bool
}

// reciprocalRankFusion fuses an arbitrary number of ranked lists into
// a single ranking. Order of the input lists does not change the
// fused score; it only controls which list seeds the metadata for a
// candidate seen in multiple lists. Later lists fill in blanks but
// never overwrite non-empty fields (the one-way merge rule from the
// spec).
//
// Ties on score are broken by path ascending so the output is stable
// across runs with identical inputs.
func reciprocalRankFusion(lists [][]rrfCandidate, k int) []RetrievedChunk {
	candidates := reciprocalRankFusionCandidates(lists, k)
	out := make([]RetrievedChunk, 0, len(candidates))
	for _, c := range candidates {
		chunk := RetrievedChunk{
			ChunkID:    c.id,
			DocumentID: c.id,
			Path:       c.path,
			Score:      c.score,
			Text:       c.content,
			Title:      c.title,
			Summary:    c.summary,
		}
		if c.haveBM25Rank {
			chunk.BM25Rank = c.bm25Rank
		}
		if c.haveVectorSim {
			chunk.VectorSimilarity = c.vectorSimilarity
		}
		out = append(out, chunk)
	}
	return out
}

type fusedCandidate struct {
	id               string
	path             string
	title            string
	summary          string
	content          string
	bm25Rank         int
	haveBM25Rank     bool
	vectorSimilarity float64
	haveVectorSim    bool
	score            float64
}

func reciprocalRankFusionCandidates(lists [][]rrfCandidate, k int) []fusedCandidate {
	safeK := k
	if safeK <= 0 {
		safeK = RRFDefaultK
	}

	type bucket struct {
		id               string
		path             string
		title            string
		summary          string
		content          string
		bm25Rank         int
		haveBM25Rank     bool
		vectorSimilarity float64
		haveVectorSim    bool
		score            float64
	}

	buckets := make(map[string]*bucket)
	order := make([]string, 0)

	for _, list := range lists {
		for rank, c := range list {
			if c.id == "" {
				continue
			}
			contribution := 1.0 / float64(safeK+rank+1)
			existing, ok := buckets[c.id]
			if !ok {
				b := &bucket{
					id:      c.id,
					path:    c.path,
					title:   c.title,
					summary: c.summary,
					content: c.content,
					score:   contribution,
				}
				if c.haveBM25Rank {
					b.bm25Rank = c.bm25Rank
					b.haveBM25Rank = true
				}
				if c.haveVectorSim {
					b.vectorSimilarity = c.vectorSimilarity
					b.haveVectorSim = true
				}
				buckets[c.id] = b
				order = append(order, c.id)
				continue
			}

			// One-way fill: later lists seed only when the first
			// list left the field empty. BM25 tends to carry full
			// title + summary; vector hits can arrive bare.
			if existing.title == "" && c.title != "" {
				existing.title = c.title
			}
			if existing.summary == "" && c.summary != "" {
				existing.summary = c.summary
			}
			if existing.content == "" && c.content != "" {
				existing.content = c.content
			}
			if !existing.haveBM25Rank && c.haveBM25Rank {
				existing.bm25Rank = c.bm25Rank
				existing.haveBM25Rank = true
			}
			if !existing.haveVectorSim && c.haveVectorSim {
				existing.vectorSimilarity = c.vectorSimilarity
				existing.haveVectorSim = true
			}
			existing.score += contribution
		}
	}

	out := make([]fusedCandidate, 0, len(order))
	for _, id := range order {
		b := buckets[id]
		out = append(out, fusedCandidate{
			id:               b.id,
			path:             b.path,
			title:            b.title,
			summary:          b.summary,
			content:          b.content,
			bm25Rank:         b.bm25Rank,
			haveBM25Rank:     b.haveBM25Rank,
			vectorSimilarity: b.vectorSimilarity,
			haveVectorSim:    b.haveVectorSim,
			score:            b.score,
		})
	}

	sort.SliceStable(out, func(i, j int) bool {
		if out[i].score != out[j].score {
			return out[i].score > out[j].score
		}
		return out[i].path < out[j].path
	})
	return out
}
